### PR TITLE
feat(18530): hide reference and selected area on layers panel if they are empty

### PR DESCRIPTION
**Pull request by Natallia-Harshunova**
https://kontur.fibery.io/Tasks/Task/Hide-Reference-and-Selected-area-on-Layers-panel-if-they-are-empty-18530

## Summary by CodeRabbit

- **New Features**
  - Enhanced layer management with filtering capabilities in the LayersTree and PanelContent components.
  
- **Refactor**
  - Updated atom creation and usage in the LayersTree to improve performance and maintainability.

- **Bug Fixes**
  - Improved logic for layer filtering based on specific conditions to ensure accurate data display in the PanelContent component.

- **Miscellaneous**
  - Updated `useAtom` function imports for better compatibility and performance.

### CODE REVIEW

File: src/features/legend_panel/components/PanelContent/PanelContent.tsx [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
+import { focusedGeometryAtom } from '~core/focused_geometry/model';
+import { referenceAreaAtom } from '~core/shared_state/referenceArea';
+import { REFERENCE_AREA_LOGICAL_LAYER_ID } from '~features/reference_area/constants';
 import { LegendsList } from '../LegendsList/LegendsList';
 import s from './PanelContent.module.css';
 
 export function PanelContent() {
   const [layers] = useAtom(mountedLayersAtom);
-  const layersAtoms = useMemo(() => Array.from(layers.values()), [layers]);
+  const [focusedGeometry] = useAtom(focusedGeometryAtom);
```
Comment (albaranau) [inline src/features/legend_panel/components/PanelContent/PanelContent.tsx]: You can use `focusedGeometryAtom.v3atom` to avoid importing `@reatom/react-v2`

**Review by albaranau — APPROVED** [review]

File: src/core/logical_layers/atoms/layersTree/layersTree.ts [hunk 2]
```diff
@@ -1,27 +1,32 @@
...
+  const hierarchy = ctx.spy(logicalLayersHierarchyAtom.v3atom);
+  const categoriesSettings = ctx.spy(layersCategoriesSettingsAtom.v3atom);
+  const groupsSettings = ctx.spy(layersGroupsSettingsAtom.v3atom);
+  const focusedGeometry = ctx.spy(focusedGeometryAtom.v3atom);
+  const referenceGeometry = ctx.spy(referenceAreaAtom);
+
+  const layers = Object.values(hierarchy).filter((layer) => {
+    if (layer.id === FOCUSED_GEOMETRY_LOGICAL_LAYER_ID) return focusedGeometry;
+    if (layer.id === REFERENCE_AREA_LOGICAL_LAYER_ID) return referenceGeometry;
+    else return true;
```
Comment (dqunbp) [inline src/core/logical_layers/atoms/layersTree/layersTree.ts:23]: It looks like `else` is not required here.

File: src/features/legend_panel/components/PanelContent/PanelContent.tsx [hunk 3]
```diff
@@ -1,12 +1,25 @@
...
 export function PanelContent() {
-  const [layers] = useAtom(mountedLayersAtom);
-  const layersAtoms = useMemo(() => Array.from(layers.values()), [layers]);
+  const [layers] = useAtom(mountedLayersAtom.v3atom);
+  const [focusedGeometry] = useAtom(focusedGeometryAtom.v3atom);
+  const [referenceAreaGeometry] = useAtom(referenceAreaAtom);
+
+  const layersAtoms = useMemo(() => {
+    return Array.from(layers.values()).filter((layer) => {
+      if (layer.id === FOCUSED_GEOMETRY_LOGICAL_LAYER_ID) return focusedGeometry;
```
Comment (dqunbp) [inline src/features/legend_panel/components/PanelContent/PanelContent.tsx:18]: Can we somehow avoid the code duplication?

File: src/features/legend_panel/components/PanelContent/PanelContent.tsx [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
+      if (layer.id === REFERENCE_AREA_LOGICAL_LAYER_ID) return referenceAreaGeometry;
+      else return true;
```
Comment (dqunbp) [inline src/features/legend_panel/components/PanelContent/PanelContent.tsx:20]: Can we somehow avoid the code duplication?

File: src/core/logical_layers/atoms/layersTree/layersTree.ts (hunk 2 above)
Comment (Natallia-Harshunova) [reply src/core/logical_layers/atoms/layersTree/layersTree.ts:23]: No, we need to use else clause here to keep other layers on panels, while there is a requirement in story to hide selected/referenced area if they are empty.

File: src/features/legend_panel/components/PanelContent/PanelContent.tsx (hunk 4 above)
Comment (Natallia-Harshunova) [reply src/features/legend_panel/components/PanelContent/PanelContent.tsx:20]: Yeah, that's a good idea. We can avoid duplication by defining a filter util  function. Maybe you can help me and suggest, where I can store this helper, cause its used in 2 different features: `legend_panel` and `layers_panel`.

File: src/core/logical_layers/atoms/layersTree/layersTree.ts (hunk 2 above)
Comment (dqunbp) [reply src/core/logical_layers/atoms/layersTree/layersTree.ts:23]: I mean `return true` without else.

File: src/features/legend_panel/components/PanelContent/PanelContent.tsx (hunk 4 above)
Comment (dqunbp) [reply src/features/legend_panel/components/PanelContent/PanelContent.tsx:20]: Somewhere in the `core`, all common code is located in the `core` directory.

**Review by dqunbp — APPROVED** [review]

**Review by albaranau — APPROVED** [review]

**Pull request merged**